dev/src/bundle.ts

```diff
@@ -0,0 +1,225 @@
+
+  // DocumentSnapshot and QuerySnapshot added to the bundle.
+  private docSnaps: DocumentSnapshot[] = [];
+  private querySnaps: Map<string, QuerySnapshot> = new Map<
+    string,
+    QuerySnapshot
+  >();
+
+  // Resulting documents for the bundle.
+  private documents: Map<string, BundledDocument> = new Map();
+  // Named queries saved in the bundle.
+  private namedQueries: Map<string, firestore.INamedQuery> = new Map();
+
+  // The latest read time among all bundled documents and queries.
+  private latestReadTime: Timestamp = Timestamp.fromProto({
+    seconds: 0,
+    nanos: 0,
+  });
+
+  constructor(private id: string) {}
+
+  add(doc: DocumentReference): BundleBuilder;
+  add(docSnap: DocumentSnapshot): BundleBuilder;
+  add(queryName: string, querySnap: QuerySnapshot): BundleBuilder;
+  add(queryName: string, query: Query): BundleBuilder;
+  add(
+    docOrName: DocumentReference | DocumentSnapshot | string,
+    query?: QuerySnapshot | Query
+  ) {
+    if (docOrName instanceof DocumentReference) {
+      this.docRefs.push(docOrName);
+    } else if (docOrName instanceof DocumentSnapshot) {
+      this.docSnaps.push(docOrName);
+    } else if (query instanceof Query) {
+      this.queries.set(docOrName, query);
+    } else if (query instanceof QuerySnapshot) {
+      this.querySnaps.set(docOrName, query);
+    }
+
+    return this;
+  }
+
+  private addBundledDocument(snap: DocumentSnapshot) {
+    if (snap.exists) {
+      const docProto = snap.toDocumentProto();
+      if (
+        !this.documents.has(snap.id) ||
+        Timestamp.fromProto(this.documents.get(snap.id)!.metadata.readTime!) <
+          snap.readTime
+      ) {
+        this.documents.set(snap.id, {
+          document: docProto,
+          metadata: {
+            documentKey: docProto.name,
+            readTime: snap.readTime.toProto().timestampValue,
+          },
+        });
+      }
+      if (snap.readTime > this.latestReadTime) {
+        this.latestReadTime = snap.readTime;
+      }
+    }
+  }
+
+  private addNamedQuery(name: string, querySnap: QuerySnapshot) {
+    this.namedQueries.set(name, {
+      name,
+      bundledQuery: querySnap.query.toBundledQuery(),
+      readTime: querySnap.readTime.toProto().timestampValue,
+    });
+
+    for (const snap of querySnap.docs) {
+      this.addBundledDocument(snap);
+    }
+
+    if (querySnap.readTime > this.latestReadTime) {
+      this.latestReadTime = querySnap.readTime;
+    }
+  }
+
+  private pushLengthPrefixedString(readable: Readable, payload: string) {
+    const buffer = Buffer.from(payload, 'utf-8');
+    const lengthBuffer = Buffer.from(buffer.length.toString());
+
+    readable.push(Buffer.concat([lengthBuffer, buffer]));
+  }
+
+  stream(): NodeJS.ReadableStream {
+    assert(
+      this.docRefs.length +
+        this.docSnaps.length +
+        this.queries.size +
+        this.querySnaps.size >
+        0,
+      'Nothing is added to the bundle.'
+    );
```

Comment:
If you do want throw an error here, it should be validation error rather than an assert. This is caused by the user and not by us and our SDK should only trigger assertions for problems that the SDK created.

I do however think that it would be more user friendly to just return an empty bundle, as this behavior is less surprising than an exception that might only occur under certain usage scenarios that otherwise are valid.